File: app/src/main/java/io/oldering/tvfoot/red/matches/displayable/MatchRowDisplayable.java
package io.oldering.tvfoot.red.matches.displayable;

import android.support.annotation.VisibleForTesting;
import com.google.auto.value.AutoValue;
import io.oldering.tvfoot.red.data.entity.Broadcaster;
import io.oldering.tvfoot.red.data.entity.Competition;
import io.oldering.tvfoot.red.data.entity.Match;
import io.oldering.tvfoot.red.data.entity.Team;
import io.oldering.tvfoot.red.util.StringUtils;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;
import javax.annotation.Nullable;

import static io.oldering.tvfoot.red.util.TimeConstants.ONE_MATCH_TIME_IN_MILLIS;

@AutoValue public abstract class MatchRowDisplayable implements MatchesItemDisplayable {
  private static SimpleDateFormat shortDateFormat = new SimpleDateFormat("HH:mm", Locale.FRANCE);
  static SimpleDateFormat mediumDateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.FRANCE);
  private static SimpleDateFormat fullTextDateFormat =
      new SimpleDateFormat("EEEE dd MMMM yyyy à HH'h'mm", Locale.FRANCE);

  public abstract String headerKey();

  public abstract String startTime();

  public abstract List<BroadcasterRowDisplayable> broadcasters();

  public abstract String headline();

  public abstract String competition();

  public abstract String matchDay();

  public abstract boolean live();

  public abstract String startTimeInText();

  public abstract String homeTeamDrawableName();

  public abstract String awayTeamDrawableName();

  public abstract String location();

  public abstract String matchId();

  @VisibleForTesting public static MatchRowDisplayable fromMatch(Match match) {
    return new AutoValue_MatchRowDisplayable( //
        parseHeaderKey(match.startAt()), //
        parseStartTime(match.startAt()), //
        parseBroadcasters(match.broadcasters()), //
        parseHeadLine(match.homeTeam(), match.awayTeam(), match.label()), //
        parseCompetition(match.competition()), //
        parseMatchDay(match.label(), match.matchDay()), //
        isMatchLive(match.startAt()), //
        parseStartTimeInText(match.startAt()), //
        parseHomeTeamDrawableName(match.homeTeam()), //
        parseAwayTeamDrawableName(match.awayTeam()), //
        parseLocation(match), match.id());
  }

  public static List<MatchRowDisplayable> fromMatches(List<Match> matches) {
    List<MatchRowDisplayable> matchRowDisplayables = new ArrayList<>(matches.size());
    for (Match match : matches) {
      matchRowDisplayables.add(fromMatch(match));
    }
    return matchRowDisplayables;
  }

  private static String parseHeaderKey(Date startAt) {
    mediumDateFormat.setTimeZone(TimeZone.getDefault());
    return mediumDateFormat.format(startAt);
  }

  private static String parseStartTime(Date startAt) {
    shortDateFormat.setTimeZone(TimeZone.getDefault());
    return shortDateFormat.format(startAt);
  }

  private static List<BroadcasterRowDisplayable> parseBroadcasters(
      @Nullable List<Broadcaster> broadcasters) {
    if (broadcasters == null) {
      return new ArrayList<>();
    }

    List<BroadcasterRowDisplayable> broadcastersVM = new ArrayList<>(broadcasters.size());
    for (Broadcaster broadcaster : broadcasters) {
      broadcastersVM.add(BroadcasterRowDisplayable.builder().code(broadcaster.code()).build());
    }
    return broadcastersVM;
  }

  private static String parseHeadLine(Team homeTeam, Team awayTeam, @Nullable String matchLabel) {
    if (homeTeam.isEmpty() || awayTeam.isEmpty()) {
      return String.valueOf(matchLabel);
    }
    return String.valueOf(homeTeam.name()).toUpperCase() + " - " + String.valueOf(awayTeam.name())
        .toUpperCase();
  }

  private static String parseCompetition(Competition competition) {
    return competition.name();
  }

  private static String parseMatchDay(@Nullable String matchLabel, @Nullable String matchDay) {
    if (matchLabel != null && !matchLabel.trim().isEmpty()) {
      return matchLabel;
    } else {
      return "J. " + matchDay;
    }
  }

  private static boolean isMatchLive(Date startAt) {
    long now = Calendar.getInstance().getTimeInMillis();
    long startTimeInMillis = startAt.getTime();
    return now >= startTimeInMillis && now <= startTimeInMillis + ONE_MATCH_TIME_IN_MILLIS;
  }

  private static String parseStartTimeInText(Date startAt) {
    fullTextDateFormat.setTimeZone(TimeZone.getDefault());
    return StringUtils.capitalize(fullTextDateFormat.format(startAt));
  }

  private static String parseHomeTeamDrawableName(Team homeTeam) {
    String code = homeTeam.code();
    if (code != null) {
      return code;
    }
    return Team.DEFAULT_CODE;
  }

  private static String parseAwayTeamDrawableName(Team awayTeam) {
    String code = awayTeam.code();
    if (code != null) {
      return code;
    }
    return Team.DEFAULT_CODE;
  }

  private static String parseLocation(Match match) {
    return String.valueOf(match.place());
  }

  @Override public boolean isSameAs(MatchesItemDisplayable newItem) {
    return newItem instanceof MatchRowDisplayable && this.matchId()
        .equals(((MatchRowDisplayable) newItem).matchId());
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/displayable/MatchRowDisplayable.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable

import com.benoitquenaudon.tvfoot.red.app.data.entity.Broadcaster
import com.benoitquenaudon.tvfoot.red.app.data.entity.Competition
import com.benoitquenaudon.tvfoot.red.app.data.entity.Match
import com.benoitquenaudon.tvfoot.red.app.data.entity.Team
import com.benoitquenaudon.tvfoot.red.util.MatchId
import com.benoitquenaudon.tvfoot.red.util.WillBeNotified
import com.benoitquenaudon.tvfoot.red.util.formatter
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Calendar
import java.util.Date
import java.util.concurrent.TimeUnit

@Suppress("DataClassPrivateConstructor")
data class MatchRowDisplayable private constructor(
    val headerKey: String,
    val startTime: String,
    val broadcasters: List<BroadcasterRowDisplayable>,
    val headline: String,
    val competition: String,
    val competitionCode: String,
    val matchDay: String?,
    val live: Boolean,
    val homeTeamDrawableName: String,
    val awayTeamDrawableName: String,
    val location: String,
    val matchId: String,
    val tags: List<String>,
    val homeTeam: TeamRowDisplayable,
    val awayTeam: TeamRowDisplayable,
    val willBeNotified: Boolean = false,
    val startAt: Long
) : MatchesItemDisplayable, Comparable<MatchRowDisplayable> {
  override fun compareTo(other: MatchRowDisplayable): Int {
    return startAt.compareTo(other.startAt)
  }

  override fun isSameAs(other: MatchesItemDisplayable): Boolean {
    return other is MatchRowDisplayable && this.matchId == other.matchId
  }

  companion object Factory {
    private fun fromMatch(
        match: Match,
        willBeNotified: WillBeNotified
    ): MatchRowDisplayable {
      return MatchRowDisplayable(
          headerKey = parseHeaderKey(match.startAt),
          startTime = parseStartTime(match.startAt),
          broadcasters = parseBroadcasters(match.broadcasters),
          headline = parseHeadLine(match.homeTeam, match.awayTeam, match.label),
          competition = parseCompetition(match.competition),
          competitionCode = parseCompetitionCode(match.competition),
          matchDay = parseMatchDay(match.label, match.matchDay),
          live = isMatchLive(match.startAt),
          homeTeamDrawableName = parseHomeTeamDrawableName(match.homeTeam),
          awayTeamDrawableName = parseAwayTeamDrawableName(match.awayTeam),
          location = parseLocation(match),
          matchId = match.id,
          tags = parseTags(match),
          homeTeam = TeamRowDisplayable(
              match.homeTeam.name,
              match.homeTeam.logoPath,
              match.homeTeam.code),
          awayTeam = TeamRowDisplayable(
              match.awayTeam.name,
              match.awayTeam.logoPath,
              match.awayTeam.code),
          willBeNotified = willBeNotified,
          startAt = match.startAt.time
      )
    }

    fun fromMatches(
        matches: List<Match>,
        willBeNotifiedPairs: Map<MatchId, WillBeNotified>
    ): List<MatchRowDisplayable> {
      return matches.map {
        this.fromMatch(it, willBeNotifiedPairs.getOrElse(it.id, { false }))
      }
    }
  }
}

private val shortDateFormat: SimpleDateFormat by formatter("HH:mm")
val mediumDateFormat: SimpleDateFormat by formatter("yyyy-MM-dd")


private fun parseHeaderKey(startAt: Date): String = mediumDateFormat.format(startAt)

private fun parseStartTime(startAt: Date): String = shortDateFormat.format(startAt)

private fun parseBroadcasters(
    broadcasters: List<Broadcaster>?): List<BroadcasterRowDisplayable> {
  if (broadcasters == null) {
    return ArrayList()
  }

  return broadcasters
      .filter { it.name != null }
      .map { BroadcasterRowDisplayable(it.name!!, it.code) }
}

private fun parseHeadLine(homeTeam: Team, awayTeam: Team, matchLabel: String?): String {
  if (homeTeam.name.isNullOrEmpty() || awayTeam.name.isNullOrEmpty()) {
    return matchLabel ?: "undefined"
  }

  return homeTeam.name.toString().toUpperCase() +
      " - " +
      awayTeam.name.toString().toUpperCase()
}

private fun parseCompetition(competition: Competition): String = competition.name

private fun parseCompetitionCode(competition: Competition): String = competition.code

private fun parseMatchDay(matchLabel: String?, matchDay: String?): String? {
  if (matchLabel.isNullOrEmpty()) {
    if (matchDay.isNullOrEmpty()) return null // happens sometimes...

    val dayAsInt: Int? = matchDay?.toIntOrNull()
    if (dayAsInt == null) {
      return checkNotNull(matchDay)
    } else {
      return "J. $matchDay"
    }
  } else {
    return checkNotNull(matchLabel)
  }
}

private fun isMatchLive(startAt: Date): Boolean {
  val now = Calendar.getInstance().timeInMillis
  val startTimeInMillis = startAt.time

  return now in startTimeInMillis..(startTimeInMillis + TimeUnit.MINUTES.toMillis(105))
}

private fun parseHomeTeamDrawableName(homeTeam: Team): String = homeTeam.code ?: Team.DEFAULT_CODE

private fun parseAwayTeamDrawableName(awayTeam: Team): String = awayTeam.code ?: Team.DEFAULT_CODE

private fun parseLocation(match: Match): String = match.place.toString()

private fun parseTags(match: Match): List<String> =
    match.tags?.map(String::toLowerCase) ?: emptyList()